Document Getting started procedure for Argo CD Rollouts via OpenShift GitOps (#78060)
openshift/openshift-docs: modules/gitops-deploying-a-rollout.adoc

// Module included in the following assemblies:
//
// * argo_rollouts/getting-started-with-argo-rollouts.adoc

:_mod-docs-content-type: PROCEDURE
[id="gitops-deploying-a-rollout_{context}"]
= Deploying a rollout

> As a cluster administrator, you can configure Argo Rollouts to progressively route a subset of user traffic to a new application version. This allows you to test if the application is deployed and working.

Reviewer: ```suggestion
As a cluster administrator, you can configure Argo Rollouts to progressively route a subset of user traffic to a new application version. Then you can test whether the application is deployed and working.
```
Use active voice, not passive voice.

> The following example procedure creates a `rollouts-demo` rollout and service. It then routes 20% of traffic to a canary version of the application, followed by multiple manual promotions, until the entire traffic is routed to the new application version.

Author: @jgwest Perfect! I have addressed it.

> The following example procedure creates a `rollouts-demo` rollout and service. It then routes 20% of traffic to a canary version of the application, waits for a manual promotion, and then performs multiple automated promotions until it routes the entire traffic to the new application version.

Reviewer: ```suggestion
The following example procedure creates a `rollouts-demo` rollout and service. It then routes 20% of traffic to a canary version of the application, waits for a manual promotion, and then performs multiple automated promotions until it routes the entire traffic to the new application version.
```
I don't know what "it" is here. Do we mean the rollout service?

Author: No, the rollout. I have addressed it. thanks!

.Procedure 

. In the *Administrator* perspective of the web console, click *Operators* -> *Installed Operators* -> *{gitops-title}* -> *Rollout*.

> . Create or select the project where you want to create and configure a `Rollout` custom resource (CR) from the *Project* drop-down menu.

Reviewer: ```suggestion
. Create or select the project in which you want to create and configure a `Rollout` custom resource (CR) from the *Project* drop-down menu.
```

. Click *Create Rollout* and enter the following configuration in the YAML view:
+
[source,yaml]
----
apiVersion: argoproj.io/v1alpha1
kind: Rollout
metadata:
  name: rollouts-demo
spec:
  replicas: 5
  strategy:
    canary: #<1>        
      steps: #<2>
      - setWeight: 20 #<3>       
      - pause: {}  #<4>         
      - setWeight: 40 
      - pause: {duration: 45}  #<5>     
      - setWeight: 60
      - pause: {duration: 20}       
      - setWeight: 80
      - pause: {duration: 10}
  revisionHistoryLimit: 2
  selector:
    matchLabels:
      app: rollouts-demo
  template: #<6>
    metadata:
      labels:
        app: rollouts-demo
    spec:
      containers:
      - name: rollouts-demo
        image: argoproj/rollouts-demo:blue
        ports:
        - name: http
          containerPort: 8080
          protocol: TCP
        resources:
          requests:
            memory: 32Mi
            cpu: 5m
----

> <1> The deployment strategy that the rollout must use.

Reviewer: What are valid possible values here and in the following steps?

Author: the value is canary that we must use here as the deployment strategy. Already it is mentioned.

<2> Specify the steps for the rollout. This example gradually routes 20%, 40%, 60%, and 80% of traffic to the canary version.
<3> The percentage of traffic that must be directed to the canary version. A value of 20 means that 20% of traffic is directed to the canary version.

> <4> Specify to the Argo Rollouts controller to pause indefinitely until it finds a request for promotion specified in the subsequent step.

Author: @jgwest Agree. Done. Thanks for the input and explanation. This is helpful!

> <5> Specify to the Argo Rollouts controller to pause for a duration of 45 seconds. You can set the duration value in seconds, minutes, or hours. If no value is specified, the duration value defaults to seconds.

Reviewer: What do various values look like? If 45 is seconds, do I use 120 or 2:00 for two minutes? What do I use for hours?

<6> Specifies the pods that are to be created.

. Click *Create*.
+
[NOTE]
====

> To ensure that the Rollout becomes available quickly on creation, the Argo Rollouts controller automatically treats the `argoproj/rollouts-demo:blue` initial container image specified in the `.spec.template.spec.containers.image` field as a stable version. In the initial instance, the creation of the `Rollout` resource routes all of the traffic towards the stable version of the application and skips the part where the traffic is sent to the canary version. However, for all subsequent application upgrades with the modifications to the `.spec.template.spec.containers.image` field, the Argo Rollouts controller performs the canary steps, as usual.

Reviewer: ```suggestion
To ensure that the rollout becomes available quickly on creation, the Argo Rollouts controller automatically treats the `argoproj/rollouts-demo:blue` initial container image specified in the `.spec.template.spec.containers.image` field as a stable version. In the initial instance, the creation of the `Rollout` resource routes all of the traffic towards the stable version of the application and skips the part where the traffic is sent to the canary version. However, for all subsequent application upgrades with the modifications to the `.spec.template.spec.containers.image` field, the Argo Rollouts controller performs the canary steps, as usual.
```

----
<1> Specify the namespace where the `Rollout` resource is defined.
+
.Sample output
[source,terminal]
----
Name:            rollouts-demo
Namespace:       spring-petclinic
Status:          ✔ Healthy
Strategy:        Canary
  Step:          8/8
  SetWeight:     100
  ActualWeight:  100
Images:          argoproj/rollouts-demo:blue (stable)
Replicas:
  Desired:       5
  Current:       5
  Updated:       5
  Ready:         5
  Available:     5

NAME                                       KIND        STATUS     AGE    INFO
⟳ rollouts-demo                            Rollout     ✔ Healthy  4m50s  
└──# revision:1                                                          
   └──⧉ rollouts-demo-687d76d795           ReplicaSet  ✔ Healthy  4m50s  stable
      ├──□ rollouts-demo-687d76d795-75k57  Pod         ✔ Running  4m49s  ready:1/1
      ├──□ rollouts-demo-687d76d795-bv5zf  Pod         ✔ Running  4m49s  ready:1/1
      ├──□ rollouts-demo-687d76d795-jsxg8  Pod         ✔ Running  4m49s  ready:1/1
      ├──□ rollouts-demo-687d76d795-rsgtv  Pod         ✔ Running  4m49s  ready:1/1
      └──□ rollouts-demo-687d76d795-xrmrj  Pod         ✔ Running  4m49s  ready:1/1
----
+

> After the Rollout has been created, you can verify that the *Status* field of the Rollout shows *Phase: Healthy*.

Reviewer: ```suggestion
After the rollout has been created, you can verify that the *Status* field of the rollout shows *Phase: Healthy*.
```

> . In the *Rollout* tab, under the *Rollouts* section, verify that the *Status* field of the `rollouts-demo` rollout shows as *Phase: Available*.

Author: @jgwest Good catch, again! Thanks for the screenshot and for going the extra mile to verify this! Really appreciate!

+
[TIP]
====

> Alternatively, you can verify that the rollout is healthy by running hte following command:

Author: @jgwest Good catch! Thanks again :)
